### PR TITLE
[SPARK-30762] Add dtype=float32 support to vector_to_array UDF

### PR DESCRIPTION
**Pull request by liangz1**
### What changes were proposed in this pull request?
In this PR, we add a parameter in the python function vector_to_array(col) that allows converting to a column of arrays of Float (32bits) in scala, which would be mapped to a numpy array of dtype=float32.

### Why are the changes needed?
In the downstream ML training, using float32 instead of float64 (default) would allow a larger batch size, i.e., allow more data to fit in the memory.

### Does this PR introduce any user-facing change?
Yes.
Old: `vector_to_array()` only take one param
```
df.select(vector_to_array("colA"), ...)
```
New: `vector_to_array()` can take an additional optional param: `dtype` = "float32" (or "float64")
```
df.select(vector_to_array("colA", "float32"), ...)
```

### How was this patch tested?
Unit test in scala.
doctest in python.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/functions.scala [hunk 1]
```diff
@@ -39,10 +39,31 @@ object functions {
     }
   }.asNonNullable()
 
+  private val vectorToArrayFloatUdf = udf { vec: Any =>
+    vec match {
+      case v: Vector => v.toArray.map(_.toFloat)
+      case v: OldVector => v.toArray.map(_.toFloat)
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/functions.scala]: Why not merge this method into `vectorToArrayUdf` ? like:
`vectorToArrayUdf = udf { (vec: Any, type: String) => ... }`

**Comment by WeichenXu123** [comment]
ok to test

File: mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala [hunk 2]
```diff
@@ -61,5 +61,20 @@ class FunctionsSuite extends MLTest {
...
         s"but got ${valType}"))
     }
+
+    val df3 = Seq(
+      (Vectors.dense(1.0, 2.0, 3.0), OldVectors.dense(10.0, 20.0, 30.0)),
+      (Vectors.sparse(3, Seq((0, 2.0), (2, 3.0))), OldVectors.sparse(3, Seq((0, 20.0), (2, 30.0))))
+    ).toDF("vec", "oldVec")
+
+    val result3 = df3.select(
+      vector_to_array('vec, dtype = "float32"), vector_to_array('oldVec, dtype = "float32"))
```
Comment (liangz1) [inline mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]: @WeichenXu123 Do you know a good way to verify the underlying data type is float32?

Comment (WeichenXu123) [reply mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]: Check array type. Should be "Array<Float>", its corresponding dataframe type is `ArrayType(elementType=FloatType)`

Comment (WeichenXu123) [reply mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]: or check `sizeof(array[0]) === 4` ?

Comment (liangz1) [reply mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]:
```
scala> df.schema.fields(0).dataType.simpleString
res5: String = array<float>
```
What about checking this dataType.simpleString?

File: mllib/src/main/scala/org/apache/spark/ml/functions.scala [hunk 3]
```diff
@@ -21,21 +21,33 @@ import org.apache.spark.annotation.Since
...
+          case v: OldVector => v.toArray
+          case v => throw new IllegalArgumentException(
+            "function vector_to_array requires a non-null input argument and input type must be " +
+              "`org.apache.spark.ml.linalg.Vector` or `org.apache.spark.mllib.linalg.Vector`, " +
+              s"but got ${ if (v == null) "null" else v.getClass.getName }.")
+        }
+      if (dtype == "float64") {
+        new_vec
+      } else if (dtype == "float32") {
+        new_vec.map(_.toFloat)
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/functions.scala]: For sparse vector & float32 type case, here we convert it to double array and then convert it to float array.
We should do a simple optimization:
1. create float array first.
2. use `SparseVector.foreachActive` to fill this array.

File: mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala [hunk 4]
```diff
@@ -61,5 +59,34 @@ class FunctionsSuite extends MLTest {
         "`org.apache.spark.ml.linalg.Vector` or `org.apache.spark.mllib.linalg.Vector`, " +
         s"but got ${valType}"))
     }
+
+    val df3 = Seq(
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]: where is pyspark side test ?

Comment (liangz1) [reply mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]: Just added in the doctest

File: mllib/src/main/scala/org/apache/spark/ml/functions.scala (hunk 1 above)
Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/functions.scala]: @liangz1 It's my fault. We cannot combine the two UDF together. Their return type are Array<Double> and Array<Float>, they're incompatible. This leads to error in your tests.

File: mllib/src/main/scala/org/apache/spark/ml/functions.scala [hunk 5]
```diff
@@ -60,6 +58,15 @@ object functions {
    *
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/functions.scala]: Add parameter and return type description in doc.
say the return type include two case: `array<float>` or `array<double>`

**Comment by liangz1** [comment]
ok to test

File: python/pyspark/ml/functions.py [hunk 6]
```diff
@@ -20,7 +20,7 @@
 
 
 @since(3.0)
-def vector_to_array(col):
+def vector_to_array(col, dtype="float64"):
     """
     Converts a column of MLlib sparse/dense vectors into a column of dense arrays.
```
Comment (WeichenXu123) [inline python/pyspark/ml/functions.py]: Add doc for params and return type.

File: python/pyspark/ml/functions.py [hunk 7]
```diff
@@ -20,7 +20,7 @@
 
 
 @since(3.0)
```
Comment (WeichenXu123) [inline python/pyspark/ml/functions.py]: `@since(3.0.0)` ?

**Comment by WeichenXu123** [comment]
Also update the PR description.

File: mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala [hunk 8]
```diff
@@ -61,5 +59,34 @@ class FunctionsSuite extends MLTest {
...
+    val result3 = df_array_float.as[(Seq[Float], Seq[Float])].collect().toSeq
+
+    val expected3 = Seq(
+      (Seq(1.0, 2.0, 3.0), Seq(10.0, 20.0, 30.0)),
+      (Seq(2.0, 0.0, 3.0), Seq(20.0, 0.0, 30.0))
+    )
+    assert(result3 === expected3)
+
+    // Check data types are correct
+    df_array_double.schema.fields(0).dataType.simpleString == "array<double>"
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]: * there is no check here
* you can assert on the simple string of the entire schema directly:

~~~
assert(df.schema.simpleString === "...")
~~~

File: mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala [hunk 9]
```diff
@@ -61,5 +59,34 @@ class FunctionsSuite extends MLTest {
         "`org.apache.spark.ml.linalg.Vector` or `org.apache.spark.mllib.linalg.Vector`, " +
         s"but got ${valType}"))
     }
+
+    val df3 = Seq(
+      (Vectors.dense(1.0, 2.0, 3.0), OldVectors.dense(10.0, 20.0, 30.0)),
+      (Vectors.sparse(3, Seq((0, 2.0), (2, 3.0))), OldVectors.sparse(3, Seq((0, 20.0), (2, 30.0))))
+    ).toDF("vec", "oldVec")
+    val df_array_float = df3.select(
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]: use camelCase in scala. the UDF naming is an exception.

**Review by mengxr — CHANGES_REQUESTED** [review]

File: mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala [hunk 10]
```diff
@@ -77,10 +77,7 @@ class FunctionsSuite extends MLTest {
     assert(result3 === expected3)
 
     // Check data types are correct
-    df_array_double.schema.fields(0).dataType.simpleString == "array<double>"
-    df_array_double.schema.fields(1).dataType.simpleString == "array<double>"
-    df_array_float.schema.fields(0).dataType.simpleString == "array<float>"
-    df_array_float.schema.fields(1).dataType.simpleString == "array<float>"
+    dfArrayFloat.schema.simpleString == "struct<UDF(vec):array<float>,UDF(oldVec):array<float>>"
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/FunctionsSuite.scala]: @liangz1 
==> `assert(dfArrayFloat.schema.simpleString === "struct<UDF(vec):array<float>,UDF(oldVec):array<float>>")`

* add `assert`
* use `===`

**Review by WeichenXu123 — APPROVED** [review]

**Comment by WeichenXu123** [comment]
Merge to master/branch-3.0

**Pull request closed without merging**